### ISSUE TITLE
FP iter_not_returning_iterator with associated type

### ISSUE THREAD

**Issue by jendrikw** [rust-lang/rust-clippy#8225]
### Summary

I think associated types are not evaluated correctly when checking whether the return type implements `Iterator`. 

### Lint Name

iter_not_returning_iterator

### Reproducer

I tried this code: [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=%23!%5Bwarn(clippy%3A%3Aiter_not_returning_iterator)%5D%0A%0Atrait%20Foo%20%7B%0A%20%20%20%20type%20I%3B%0A%20%20%20%20fn%20iter(%26self)%20-%3E%20Self%3A%3AI%3B%0A%7D%0A%0Aimpl%20Foo%20for%20()%20%7B%0A%20%20%20%20type%20I%20%3D%20std%3A%3Avec%3A%3AIntoIter%3Cu8%3E%3B%0A%20%20%20%20fn%20iter(%26self)%20-%3E%20%3CSelf%20as%20Foo%3E%3A%3AI%20%7B%0A%20%20%20%20%20%20%20%20todo!()%0A%20%20%20%20%7D%0A%7D%0A)

```rust
#![warn(clippy::iter_not_returning_iterator)]

trait Foo {
    type I;
    fn iter(&self) -> Self::I;
}

impl Foo for () {
    type I = std::vec::IntoIter<u8>;
    fn iter(&self) -> Self::I {
        todo!()
    }
}
```

I saw this happen:

```
warning: this method is named `iter` but its return type does not implement `Iterator`
```

I expected to see this happen:

I expected the associated type to be substituted like a type alias.

### Version

```text
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-unknown-linux-gnu
release: 1.57.0
LLVM version: 13.0.0
```
